Design/Instrument-2.0/ScanningInstruments_Design.md

```diff
@@ -0,0 +1,170 @@
+
+The current plan is to access to add methods that take a detector index and a time index, in addition to the methods that take just a detector index. Accessing a scanning instrument with just a detector index would throw. This will save memory on detector metadata which would not change for different time indices, such as the monitor flag. See the end of this document for the changes to the signature of some of the public `DetectorInfo` methods.
+
+The `timeIndex` argument is optional, so algorithms that do not care about time indexing will simply return the first/only time index. `DetectorInfo` should also contain a vector of start and end times for each time index. This will mean any information stored as a time series sample log can also be obtained for the correct step, for example the monitor counts for normalisation.
+
+`DetectorInfo` will ultimately store the information for positions and rotation, so will need to do this in a way that can be accessed by both the detector index and time index. Masking would be done at the spectrum level, if it is required for a detector for some time steps, but not all. `DetectorInfo` will also hold the start and end times for each step in the scan. The objects it will need to hold for this would be:
+* `std::vector<std::vector<Kernel::V3D>> positions`
+* `std::vector<std::vector<Kernel::V3D>> rotations`
+* `std::vector<std::vector<std:pair<Kernel::DateAndTime, Kernel::DateAndTime>>> timeRanges`
+
+Appropriate typedefs should be used here for readability. All of the vector object would have the same size, the number of steps in the scan for the vector of vectors, and the number of detectors for the vector of positions.
+
+#### `SpectrumInfo`
+
+Currently [`SpectrumInfo`](https://github.com/mantidproject/mantid/blob/master/Framework/API/inc/MantidAPI/SpectrumInfo.h) has a private `getDetector` method for finding the detector for the workspace. This uses the `getSpectrum(index)` method on `MatrixWorkspace` to return the detector IDs for the spectrum.
+
+A new method should be added, for example `std::set<std::pair<detid_t, size_t>> getTimeIndexedDetectorIDs(index)`. The time indices for each detector can be stored in `SpectrumInfo` as a `std::set<std::pair<detid_t, size_t>>` within a vector of the same size as the number of workspace spectra. This will avoid the need to change methods on `MatrixWorkspace` or `ISpectrum/Spectrum` to return the set of detector IDs and time indexes for the spectrum. This is something that will help to move away from the incremental construction of `DetectorInfo`.
+
+This can be used to access methods on `DetectorInfo`, for example `m_detectorInfo.twoTheta(detIndex, timeIndex)`.
```

Comment:
Do we have examples where such an access would be used? From what I have seen in most of the Mantid algorithms, work is done either with spectra, or with detectors, so there are probably hardly any cases where we would need such a mix of `SpectrumInfo` and `DetectorInfo`.

What you are saying here is certainly right, but it should be made clear that such complex interaction is probably rarely necessary.